File: Receiver/src/Communication/Protocol/VehicleStatusSerializer.cpp
/******************************************************************************
 * Proyecto : MarioKart ESP32 RC
 * Archivo  : VehicleStatusSerializer.cpp
 * Autor    : Narciso Ivan Cisneros Acosta
 *
 * Descripción:
 * Implementación de la serialización de VehicleStatus.
 ******************************************************************************/

#include "VehicleStatusSerializer.h"

#include <cstring>

namespace MK::Protocol
{

//=============================================================================
// Serialización
//=============================================================================

bool SerializeVehicleStatus(
    const VehicleStatus& status,
    std::uint8_t* buffer,
    std::size_t length) noexcept
{
    if (buffer == nullptr)
    {
        return false;
    }

    constexpr std::size_t PayloadSize =
        sizeof(VehicleStatus);

    if (length < PayloadSize)
    {
        return false;
    }

    std::memcpy(
        buffer,
        &status,
        PayloadSize);

    return true;
}

//=============================================================================
// Deserialización
//=============================================================================

bool DeserializeVehicleStatus(
    const std::uint8_t* buffer,
    std::size_t length,
    VehicleStatus& status) noexcept
{
    if (buffer == nullptr)
    {
        return false;
    }

    constexpr std::size_t PayloadSize =
        sizeof(VehicleStatus);

    if (length < PayloadSize)
    {
        return false;
    }

    std::memcpy(
        &status,
        buffer,
        PayloadSize);

    return true;
}

} // namespace MK::Protocol

File: AndroidController/MarioKart.Android/MarioKart.Android/Communication/Protocol/VehicleStatusSerializer.cs
/******************************************************************************
 * Proyecto : MarioKart ESP32 RC
 * Archivo  : VehicleStatusSerializer.cs
 * Autor    : Narciso Ivan Cisneros Acosta
 *
 * Descripción:
 * Serializa y deserializa VehicleStatus.
 ******************************************************************************/

using MarioKart.Android.Shared;

namespace MarioKart.Android.Communication.Protocol
{
    public static class VehicleStatusSerializer
    {
        //=====================================================================
        // Deserialización
        //=====================================================================

        public static VehicleStatus Deserialize(
            byte[] payload)
        {
            if (payload == null)
            {
                return null;
            }

            //-------------------------------------------------------------
            // El payload actual contiene únicamente DrivingProfile
            //-------------------------------------------------------------

            if (payload.Length < 1)
            {
                return null;
            }

            return new VehicleStatus
            {
                DrivingProfile =
                    (DrivingProfile)payload[0]
            };
        }

        //=====================================================================
        // Serialización (por si luego la necesitamos)
        //=====================================================================

        public static byte[] Serialize(
            VehicleStatus status)
        {
            if (status == null)
            {
                return null;
            }

            return new byte[]
            {
                (byte)status.DrivingProfile
            };
        }
    }
}
